File: examples/sum.jl
# https://pauladamsmith.com/blog/2015/01/how-to-get-started-with-llvm-c-api.html

using Test

using LLVM

if length(ARGS) == 2
    x, y = parse.([Int32], ARGS[1:2])
else
    x = Int32(1)
    y = Int32(2)
end

@dispose ctx=Context() begin
    # set-up
    mod = LLVM.Module("my_module")

    param_types = [LLVM.Int32Type(), LLVM.Int32Type()]
    ret_type = LLVM.Int32Type()
    fun_type = LLVM.FunctionType(ret_type, param_types)
    sum = LLVM.Function(mod, "sum", fun_type)

    # generate IR
    @dispose builder=IRBuilder() begin
        entry = BasicBlock(sum, "entry")
        position!(builder, entry)

        tmp = add!(builder, parameters(sum)[1], parameters(sum)[2], "tmp")
        ret!(builder, tmp)

        verify(mod)
    end

    # analysis and execution
    @dispose engine=Interpreter(mod) begin
        args = [GenericValue(LLVM.Int32Type(), x),
                GenericValue(LLVM.Int32Type(), y)]

        res = LLVM.run(engine, sum, args)
        @test convert(Int, res) == x + y

        dispose.(args)
        dispose(res)
    end
end

File: examples/sum.cpp
// C++ counterpart of sum.jl

#include <string>
#include <vector>

#include <llvm/ExecutionEngine/ExecutionEngine.h>
#include <llvm/ExecutionEngine/GenericValue.h>
#include <llvm/ExecutionEngine/Interpreter.h>
#include <llvm/IR/IRBuilder.h>
#include <llvm/IR/LLVMContext.h>
#include <llvm/IR/Module.h>
#include <llvm/IR/Verifier.h>

static llvm::LLVMContext context;

int main(int argc, char *argv[]) {
  std::unique_ptr<llvm::Module> mod =
      llvm::make_unique<llvm::Module>("module", context);

  auto T_int32 = llvm::IntegerType::get(context, 32);
  auto fun_type = llvm::FunctionType::get(T_int32, {T_int32, T_int32}, false);
  auto sum = llvm::Function::Create(fun_type, llvm::Function::ExternalLinkage,
                                    "sum", mod.get());

  // generate IR
  llvm::IRBuilder<> builder(context);
  {
    auto *entry = llvm::BasicBlock::Create(context, "entry", sum);
    builder.SetInsertPoint(entry);

    auto args = sum->arg_begin();
    llvm::Value *arg1 = &(*args);
    args = std::next(args);
    llvm::Value *arg2 = &(*args);
    auto *tmp = builder.CreateAdd(arg1, arg2, "tmp");
    builder.CreateRet(tmp);

    mod->dump();
    verifyFunction(*sum);
  }

  // execute
  auto engine = llvm::EngineBuilder(std::move(mod))
                    .setEngineKind(llvm::EngineKind::Interpreter)
                    .create();
  {
    std::vector<llvm::GenericValue> args(2);
    args[0].IntVal = llvm::APInt(32, 1, /*isSigned=*/true);
    args[1].IntVal = llvm::APInt(32, 2, /*isSigned=*/true);

    auto res = engine->runFunction(sum, args);
    res.IntVal.dump();
  }

  return 0;
}
